src/rules.c

```diff
@@ -890,6 +890,8 @@ char *rules_apply(char *word_in, char *rule, int split, char *last)
 				/* HC rule: duplicate word N times */
 				unsigned char x, y;
 				POSITION(x)
+				if (x * length > RULE_WORD_SIZE && length)
+					x = RULE_WORD_SIZE / length;
 				y = x;
 				in[length*(x + 1)] = 0;
```

Comment:
I think the added two lines should use `RULE_WORD_SIZE - 1` in both places, drop the redundant check of `length` (the first condition implies `length` is non-zero), and perhaps use `x + 1` and subtract one in the assignment to `x`.